From c55a15bd5b1ca02bda7c61bbd74bc824809972cc Mon Sep 17 00:00:00 2001
From: Katherine Lai <laikatherine@google.com>
Date: Mon, 10 Mar 2025 18:53:34 +0000
Subject: [PATCH] CHROMIUM: GD: Check opcode support before reading LE features
 and states

Some old BT chipsets don't support LE_READ_LOCAL_SUPPORTED_FEATURES
and LE_READ_SUPPORTED_STATES and will crash. Wait for local supported
commands read for opcode support

Also add checks for LE_READ_BUFFER_SIZE_V2 and
LE_READ_FILTER_ACCEPT_LIST_SIZE.

Bug: 399959372, 399959372
Test: mmm packages/modules/Bluetooth
Flag: Exempt, validity check
Change-Id: Icd91e44f962134c6746ee708bc6f200395109cf0
---
 system/gd/hci/controller_impl.cc | 55 ++++++++++++++++++++++++--------
 1 file changed, 42 insertions(+), 13 deletions(-)

diff --git a/system/gd/hci/controller_impl.cc b/system/gd/hci/controller_impl.cc
index 69b2294bbe6..56b00b2284f 100644
--- a/system/gd/hci/controller_impl.cc
+++ b/system/gd/hci/controller_impl.cc
@@ -76,2 +76,7 @@ struct ControllerImpl::impl {
                     this, &ControllerImpl::impl::read_local_version_information_complete_handler));
+
+    // Wait for local supported commands read
+    std::promise<void> commands_promise;
+    auto commands_future = commands_promise.get_future();
+
     hci_->EnqueueCommand(
@@ -79,12 +84,25 @@ struct ControllerImpl::impl {
             handler_->BindOnceOn(
-                    this, &ControllerImpl::impl::read_local_supported_commands_complete_handler));
+                    this, &ControllerImpl::impl::read_local_supported_commands_complete_handler,
+                    std::move(commands_promise)));
+    commands_future.wait();
 
-    hci_->EnqueueCommand(
-            LeReadLocalSupportedFeaturesBuilder::Create(),
-            handler_->BindOnceOn(this,
-                                 &ControllerImpl::impl::le_read_local_supported_features_handler));
+    if (is_supported(OpCode::LE_READ_LOCAL_SUPPORTED_FEATURES)) {
+      hci_->EnqueueCommand(
+              LeReadLocalSupportedFeaturesBuilder::Create(),
+              handler_->BindOnceOn(
+                      this,
+                      &ControllerImpl::impl::le_read_local_supported_features_handler));
+    } else {
+      log::info("LE_READ_LOCAL_SUPPORTED_FEATURES not supported, defaulting to 0");
+      le_local_supported_features_ = 0;
+    }
 
-    hci_->EnqueueCommand(
-            LeReadSupportedStatesBuilder::Create(),
-            handler_->BindOnceOn(this, &ControllerImpl::impl::le_read_supported_states_handler));
+    if (is_supported(OpCode::LE_READ_SUPPORTED_STATES)) {
+      hci_->EnqueueCommand(
+              LeReadSupportedStatesBuilder::Create(),
+              handler_->BindOnceOn(this, &ControllerImpl::impl::le_read_supported_states_handler));
+    } else {
+      log::info("LE_READ_SUPPORTED_STATES not supported, defaulting to 0");
+      le_supported_states_ = 0;
+    }
 
@@ -129,3 +147,3 @@ struct ControllerImpl::impl {
               handler_->BindOnceOn(this, &ControllerImpl::impl::le_read_buffer_size_v2_handler));
-    } else {
+    } else if (is_supported(OpCode::LE_READ_BUFFER_SIZE_V1)) {
       hci_->EnqueueCommand(
@@ -133,2 +151,6 @@ struct ControllerImpl::impl {
               handler_->BindOnceOn(this, &ControllerImpl::impl::le_read_buffer_size_handler));
+    } else {
+      log::info("LE_READ_BUFFER_SIZE_V1 and LE_READ_BUFFER_SIZE_V2 are not supported, defaulting to 0");
+      le_buffer_size_.total_num_le_packets_ = 0;
+      le_buffer_size_.le_data_packet_length_ = 0;
     }
@@ -142,5 +164,10 @@ struct ControllerImpl::impl {
 
-    hci_->EnqueueCommand(
-            LeReadFilterAcceptListSizeBuilder::Create(),
-            handler_->BindOnceOn(this, &ControllerImpl::impl::le_read_accept_list_size_handler));
+    if (is_supported(OpCode::LE_READ_FILTER_ACCEPT_LIST_SIZE)) {
+      hci_->EnqueueCommand(
+              LeReadFilterAcceptListSizeBuilder::Create(),
+              handler_->BindOnceOn(this, &ControllerImpl::impl::le_read_accept_list_size_handler));
+    } else {
+      log::info("LE_READ_FILTER_ACCEPT_LIST_SIZE not supported, defaulting to 0");
+      le_accept_list_size_ = 0;
+    }
 
@@ -361,3 +388,4 @@ struct ControllerImpl::impl {
 
-  void read_local_supported_commands_complete_handler(CommandCompleteView view) {
+  void read_local_supported_commands_complete_handler(std::promise<void> promise,
+                                                      CommandCompleteView view) {
     auto complete_view = ReadLocalSupportedCommandsCompleteView::Create(view);
@@ -367,2 +395,3 @@ struct ControllerImpl::impl {
     local_supported_commands_ = complete_view.GetSupportedCommands();
+    promise.set_value();
   }
